tracer/src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -441,38 +421,16 @@ private void OnHostingHttpRequestInStart(object arg)
             {
                 HttpContext httpContext = requestStruct.HttpContext;
-                HttpRequest request = httpContext.Request;
-                Span span = null;
-                Scope scope = null;
                 if (shouldTrace)
                 {
                     // Use an empty resource name here, as we will likely replace it as part of the request
                     // If we don't, update it in OnHostingHttpRequestInStop or OnHostingUnhandledException
-                    scope = AspNetCoreRequestHandler.StartAspNetCorePipelineScope(tracer, httpContext, httpContext.Request, resourceName: string.Empty);
-                    span = scope.Span;
-                }
-
-                if (shouldSecure)
-                {
-                    httpContext.Response.OnCompleted(
-                        () =>
-                        {
-                            security.InstrumentationGateway.RaiseLastChanceToWriteTags(httpContext, span);
-                            return Task.CompletedTask;
-                        });
-
-                    httpContext.Response.OnStarting(
-                        () =>
-                        {
-                            // we subscribe here because in OnHostingHttpRequestInStop or HostingEndRequest it's too late,
-                            // the waf is already disposed by the registerfordispose callback, but we need to be at the end to get the real response status code
-                            security.InstrumentationGateway.RaiseRequestEnd(httpContext, request, span);
-                            return Task.CompletedTask;
-                        });
-
-                    security.InstrumentationGateway.RaiseRequestStart(httpContext, request, span);
-                    // Should we get rid of the Instrumentation Gateway, it s been making the code very cumbersome and hard to follow and an event driven model doesnt seem adapted here cause we need to get a return value from the security component to know here that we need to flush the span after blocking, hence the cumbersome action (last param here), binding parameters to avoid capturing local variables but so hard to read. It would be simpler to just call the security component with the current data, get the return, flush the span and throw if needed.
-                    security.InstrumentationGateway.RaiseBlockingOpportunity(httpContext, scope, tracer.Settings, (args) => DoBeforeRequestStops(args.Context, args.Scope, args.TracerSettings));
+                    var scope = AspNetCoreRequestHandler.StartAspNetCorePipelineScope(tracer, httpContext, httpContext.Request, resourceName: string.Empty);
+                    if (shouldSecure)
+                    {
+                        var securityTransport = new SecurityTransport(security, httpContext, scope.Span);
+                        securityTransport.ReportWafInitInfoOnce();
+                    }
```

Comment:
Should this `if` be nested inside `shouldTrace`? It was outside it before 🤔

Reply:
I think it was a mistake because if there's no active `scope` we can't fill all tags and metric tags abut waf initialization 🤔